### DISCUSSION TITLE
Apply device templates to devices after changes made to the template

### DISCUSSION THREAD

**Discussion by PieterL75** [netbox-community/netbox#8189] [General]
I ran into the fact, that I forgot to add some interfaces/psu to my device templates.
Currently, when the device template is updated, the devices using that template are not updated.

Is it a feasible feature request to create such a functionality ?
I do understand that there are quite some limitations and imperfections on this.
- what if the you delete an interface from a template
- what if the interface was already manually added to the device and the template is updated afterwards ?

any ideas ?

Pieter

**Comment by candlerb** [comment]
I wrote some code a while ago to identify and add missing components:

* a [report](https://gist.github.com/candlerb/d78886b93b36e570e1098eb1d346b76e) to highlight missing components
* a [script](https://gist.github.com/candlerb/9616db031301ec8fc737c620881d2fb8) to add missing components to all devices

The latter script runs from the CLI. I have not converted it into "Custom Script" that runs inside Netbox itself, but that would be straightforward for anyone who wants to pick it up (and also allow you to select specific devices to run against).

It won't remove any existing components, nor will it change existing components with the same name.

**Reply by PieterL75** [reply]
Awesome! This is good stuff!

Thx!

**Reply by candlerb** [reply]
I've updated this to work as a 'custom script'.  See:
https://github.com/netbox-community/reports/pull/51/files

**Reply by candlerb** [reply]
PR now merged:
https://github.com/netbox-community/reports/blob/master/reports/dcim-reports/missing_device_type_components.py (report)
https://github.com/netbox-community/reports/blob/master/scripts/add_device_type_components.py (custom script)

**Reply by themmini** [reply]
@candlerb - could you potential make the script run using the list of devices that it found from the missing report? instead of having to manually select each device?

**Reply by candlerb** [reply]
If you want to add *all* missing components to *all* devices highlighted in the report, then you can modify the script to iterate over `Device.objects.all()` instead of `data["devices"]`. But I thought it was safest to force you to select the devices of interest.

**Reply by Pifan91** [reply]
@candlerb 
Please help with  add_device_type_components.py

An exception occurred: IntegrityError: null value in column "level" of relation "dcim_modulebay" violates not-null constraint DETAIL: Failing row contains (592, 2025-02-20 09:03:29.895697+00, 2025-02-20 09:03:29.895723+00, {}, PIC 1, PIC 00000001, 1, 1, , 530, null, null, null, null, null, null).

**Reply by candlerb** [reply]
Netbox version?

The script just uses `template.instantiate(device)` to create the components. It seems that there is a "level" attribute in ModuleBay but not ModuleBayTemplate. Hence ModuleBayTemplate.instantiate() requires you to pass level=N explicitly.

I'm not sure even where the "level" comes from, although I see

```
class ModuleBay(ModularComponentModel, TrackingModelMixin, MPTTModel):
```

so perhaps it's from MPTTModel? Would level=0 be a sensible default?

To test this, you could try modifying the script like this:

```
--- a/scripts/add_device_type_components.py
+++ b/scripts/add_device_type_components.py
@@ -56,7 +56,7 @@ class AddDeviceTypeComponents(Script):
                 names = {i.name for i in getattr(device, item).all()}
                 templates = getattr(dt, templateitem).all()
                 items = [
-                    x.instantiate(device=device)
+                    (x.instantiate(device=device, level=0) if isinstance(x, ModuleBay) else x.instantiate(device=device))
                     for x in templates
                     if x.name not in names
                 ]
```

although maybe Netbox core maintainers would accept a change to Netbox core to make this simpler:

```
--- a/netbox/dcim/models/device_component_templates.py
+++ b/netbox/dcim/models/device_component_templates.py
@@ -685,11 +685,12 @@ class ModuleBayTemplate(ModularComponentTemplateModel):
         verbose_name = _('module bay template')
         verbose_name_plural = _('module bay templates')

-    def instantiate(self, **kwargs):
+    def instantiate(self, level=0, **kwargs):
         return self.component_model(
             name=self.name,
             label=self.label,
             position=self.position,
+            level=level,
             **kwargs
         )
     instantiate.do_not_call_in_templates = True
```

Aside: looks like MPTTModel was added to ModuleBay in 796b9e84 (Aug 6 2024) which means it made it into v4.1.0

**Reply by Pifan91** [reply]
> Netbox version?

v4.1.11

**Reply by candlerb** [reply]
OK, try the first patch I gave, and if it works I'll submit it as a pull request. I don't use module bays in any active installations I have access to.

**Reply by Pifan91** [reply]
> To test this, you could try modifying the script like this:

An exception occurred: TypeError: 'ModuleBay' object is not callable

**Reply by candlerb** [reply]
Maybe you didn't apply the patch correctly. The line starting "-" needs to be removed, and replaced with the line starting with "+" (but without the +)

That is:

```
                items = [
                    (x.instantiate(device=device, level=0) if isinstance(x, ModuleBay) else x.instantiate(device=device))
                    for x in templates
                    if x.name not in names
                ]
```

**Reply by Pifan91** [reply]
Yes, but it didn't help

`IntegrityError: null value in column "level" of relation "dcim_modulebay" violates not-null constraint DETAIL:`


Sorry, I don't know how to program.
just wanted to solve the problem with modulebay

**Comment by jeremystretch** [comment]
If you need to add additional interfaces to existing devices of a certain type, simply filter the devices list by type, select all, then click Add components > interfaces. This will accomplish exactly what you want while avoiding any nasty surprises.

**Reply by ravinald** [reply]
I'm not sure if those in this discussion are approaching it as I am, but I look at templates as the foundation by which a device is configured.  While your method is functionally the same, it bypasses the function of a template as I use them.  It would be great if the initial feature could exist, and to mitigate reconciliation I would be happy if was additive such that only Components that are on the template and not on the device will be added.  Nothing that already exists will be altered or deleted.

Edit: As I sit here musing on being in a similar situation I find my options are:

1. Create a new, duplicate device using the Device Template and manually adding back attributes (serial, cf, etc)
2. Add the dozen or so interface I didn't have part of the template at initial configuration with your method

I can't shake the feeling that while 2 works it isn't what I feel is the more correct method.  I thing it is, in part, that I would be doing the same task twice -- once to update the template and the second time to update all the existing devices.

**Reply by candlerb** [reply]
> only Components that are on the template and not on the device will be added. Nothing that already exists will be altered or deleted

That's exactly what the [script](https://github.com/netbox-community/customizations/blob/master/scripts/add_device_type_components.py) I linked to does.

> I thing it is, in part, that I would be doing the same task twice -- once to update the template and the second time to update all the existing devices

Not really. You do the job once on the device type ("template" as you call it), where you define things like all the interface types.  Then you just do an update operation on the devices, so that they pick up all these interfaces you've defined (you don't need to defined them again).  Doing it on the device type also means that all *future* instances of the device type will be correct.

**Reply by ravinald** [reply]
@candlerb I was replying to @jeremystretch's comment. Your script indeed does solve the feature/limitation, and thank you for it.
